### PR TITLE
General parser adf

### LINKED ISSUE #185: ADF Tape21 parser
**Is your feature request related to a problem? Please describe.**
ADF produces the TAPE21 files which are binary files with very specific formatting. Using the dmpkf utility that they have to convert it into an ASCII formatted file we could parse all of the data. The advantage of this is that it is less likely that there would be any changes to the format of the files along with it being universal for parsing things like atom tables which have a different format for nearly every time the atomic positions are listed.

**Describe alternatives you've considered**
Done locally.

### LINKED ISSUE #261: ADF parsers for new version
**Is your feature request related to a problem? Please describe.**
ADF/SCM seems to be making changes to the way their outputs are printing as they have transitioned into their new AMS driver. While there are a lot of similarities between the outputs for the new and old versions there are still changes that are not backwards compatible.

**Describe the solution you'd like**
Implement new parsers for the AMS/ADF driver. Along with this I think it's a good idea to create a parser that can handle the different modules that can be used like NMR and ADF. This would allow us to separate the parsing functions to what is pertinent to each of the modules and make the code more readable. An example of this is the current implementation of the `adf.Output.parse_atom` method as it has multiple if statements if one uses the ADF or NMR modules.

**Additional context**
Not sure if it would be better to have the classes in their own files, such as `nmr_output.py`, `adf_output.py`, and `ams_output.py`. Open to additional ideas.

**Current work**
Currently, I have a working parser that stores the module that we are working with in the `meta` class attribute and we call the methods from the parsing class.

``` python
class Output(six.with_metaclass(OutMeta, Editor)):
    def _parse_data(self, attr):
        try:
            parser = getattr(self.meta['module'], attr)
        except AttributeError as e:
            if str(e).endswith("'"+attr+"'"):
                return
        parser(self)

    def parse_atom(self):
        self._parse_data('parse_atom')

    def __init__(self, file, *args, **kwargs):
        ed = _Editor(file)
        _readf = "A D F"
        _renmr = "N M R"
        _reams = "A M S"
        found = ed.find(_readf, _renmr, _reams, keys_only=True)
        super(Output, self).__init__(file, *args, **kwargs)
        if found[_renmr]:
            self.meta.update({'module': NMR})
        elif found[_readf]:
            self.meta.update({'module': ADF})
        elif found[_reams]:
            self.meta.update({'module': AMS})
        else:
            raise ValueError("Could not identify program from SCM")

```

### PR DESCRIPTION
**Pull request by herbertludowieg**
New implementations:
- Parsing methods for new AMS driver
- Separated the NMR parsing methods from the general ADF class
- Generalized output parsing class for the entire ADF program package
- Parsing methods for an ASCII TAPE21 file

Updates:
- Updated all tests in test_output.py to reflect new changes

To do:
- Add tests for the TAPE21 parser

Will close #261 and close #185.

### CODE REVIEW

File: exatomic/adf/tape21.py [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+
+        Returns:
+            r_mass (:class:`numpy.ndarray`): Array containing the calculated reduced
+                                             masses.
+        '''
+        cols = ['dx', 'dy', 'dz']
+        # get the isotopic masses of the unique atoms
+        mapper = sym2isomass(symbol)
+        # get a list of the isotopic masses
+        mass = list(map(mapper.get, symbol))
```
Comment (avmarchenko) [inline exatomic/adf/tape21.py:100]: Consider DRY-ing up the repeated code `mapper = ...; mass = ...`. It seems like these methods could actually be standalone general functions (perhaps in core).

File: exatomic/adf/tape21.py [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+            ndisps = int(self[found[_renorm][0]+1].split()[0])
+            normalmodes = self._dfme(np.array(found[_renorm]), ndisps, idx=0)
+            calc_rmass = True
+        elif found[_recartnorm] and cart:
+            # get the non-mass-weighted normal modes and toss warning
+            ndisps = int(self[found[_recartnorm][0]+1].split()[0])
+            normalmodes = self._dfme(np.array(found[_recartnorm]), ndisps, idx=0)
+            calc_rmass = True
+        else:
+            raise Exception("Something went wrong")
```
Comment (avmarchenko) [inline exatomic/adf/tape21.py:182]: I would suggest dumping the values of `found`, `_renorm`, `_recartnorm` , etc. to make the exception message more valuable (also `ValueError`?).

File: exatomic/adf/tests/test_output.py [hunk 3]
```diff
@@ -5,10 +5,160 @@
 import pandas as pd
 from unittest import TestCase
 from exatomic.base import resource
-from exatomic.adf.output import Output
+from exatomic.adf.output import Output, ADF, AMS, NMR
 
 
+class TestNMROutput(TestCase):
```
Comment (avmarchenko) [inline exatomic/adf/tests/test_output.py:11]: Thank you for adding tests

**Review by avmarchenko — APPROVED** [review]
I added some comments that do not block, but I think would improve the code if implemented.

File: exatomic/adf/tape21.py (hunk 2 above)
Comment (herbertludowieg) [reply exatomic/adf/tape21.py]: I agree. I made a change in the latest commit to use the `MissingSection` error as that is the only time it should ever get to that else unless there's a bug somewhere.

File: exatomic/adf/tape21.py (hunk 1 above)
Comment (herbertludowieg) [reply exatomic/adf/tape21.py]: You're right! I addressed this in the latest commit.

**Comment by avmarchenko** [comment]
Thank you for these changes. I'm good to merge - let me know.

**Comment by herbertludowieg** [comment]
Thank you! This is ready to be merged!!

**Pull request merged**